[1단계 - 로또(자동)] 알렉스(이영환) 미션 제출합니다.

yxxnghwan opened the pull request:
반갑습니다. 알렉스입니다😁 잘 부탁드려요 또링!

이번 미션 진행하면서 궁금했던 점 남깁니다.

1. 이번 미션에서 단순히 InputView, OutputView 두 가지 View로 모든 UI로직을 담기엔 한 파일에 코드가 길어지고, 한 클래스가 여러 기능을 해야한다고 생각하게 되어서 각 기능마다 View 클래스를 따로 만들었는데요, View 클래스들의 공통적인 부분을 묶어 인터페이스로 만들었는데, 막상 인터페이스의 주요 기능인 다형성을 사용하지는 않았습니다. 이러한 구조의 경우 오버엔지니어링이었을까요?🤔
2. View에서 모델의 비즈니스 로직을 노출시키지 않으려고 DTO를 넘기는 것으로 알고 있습니다. 근데 로직을 가지지 않고 상수처럼 사용하는 enum도 DTO 써서 View에 전달하는게 맞는지 궁금합니다.(처음엔 DTO없이 View에 enum을 넘기도록 구현했었으나, enum이 가지는 정보를 축소하면서 DTO를 생성해서 넘기는 구조로 변경했습니다.)
3. 전략패턴을 최초로 주입하는 위치가 궁금합니다. 현재는 메인메서드부터 전략패턴을 주입해서 메인메서드->컨트롤러->도메인 이렇게 전달하면서 전략 패턴을 사용중인데요, 어느 부분에서부터 전략 패턴을 정하고 파라미터를 넘겨주지 않아도 되는지 궁금합니다!

페어 프로그래밍을 진행하는 동안 코드 윗 미 환경에 제대로 익숙하지 않아서 
중간에 몇몇 파일이 커밋에 포함되지 않고 있었는데, 그걸 나중에 알게 되었습니다...ㅠㅠ
다음번엔 신경 써서 커밋 체크 하겠습니다.

그 외에도 미숙한 부분이 많이 있었다고 생각되는데, 또링이 지적해주시면 가슴에 새기고 고쳐오겠습니다!

yxxnghwan committed:
docs(README): 구현할 기능 목록 작성

yxxnghwan committed:
docs(README): 구입 금액이 정수인지 검증 추가

yxxnghwan committed:
feat: 구입 금액이 정수인지 검증

yxxnghwan committed:
feat: 구입 금액이 1000의 배수인지 검증

yxxnghwan committed:
docs(README): 불필요한 기능 목록 삭제

yxxnghwan committed:
docs(README): 목록 추가 및 수정

yxxnghwan committed:
feat: 로또 번호 생성, 로또 티켓 생성

yxxnghwan committed:
feat: 로또 티켓 저장

yxxnghwan committed:
docs(README): 목록 추가 및 구현 목록 체크

yxxnghwan committed:
feat: 당첨 번호, 보너스 볼 번호 검사 기능 생성

yxxnghwan committed:
feat: 전략패턴 수정 및 로또 당첨 결과 기능 구현

전략패턴 번호생성을 int -> List<Integer>로 한번에 여러개 생성하도록 변경
생성된 번호랑 사용자가 입력한 번호랑 비교하여 당첨 카운팅하여 결과 반환

yxxnghwan committed:
docs(README): 구현 완료한 기능목록 체크

yxxnghwan committed:
feat: 구입금액 입력 요청 및 사용자 입력

yxxnghwan committed:
feat: 생성된 로또 티켓 출력

yxxnghwan committed:
feat: 당첨 번호 입력 기능 구현

yxxnghwan committed:
refactor(validator): WinningNumberValidator -> LottoNumberValidator로 변경

yxxnghwan committed:
feat: 보너스 볼 입력 구현

yxxnghwan committed:
feat: 당첨 통계 출력

당첨 통계 출력
수익률 출력

yxxnghwan committed:
feat: LottoGame 생성자 시그니처 변경, LottoTickets 일급 컬렉션 정의, 메인 로직 구현

yxxnghwan committed:
refactor: 프로덕션 코드 전체 리팩토링

yxxnghwan committed:
test: 로또 번호 검증 테스트

yxxnghwan committed:
chore: 안쓰는 클래스 삭제

yxxnghwan committed:
test: LottoController 테스트 케이스 작성

yxxnghwan committed:
test: LottoTickets 테스트 케이스 작성

특정 파일들 커밋되지 않던 오류 수정...

yxxnghwan committed:
refactor: 2등 당첨을 판단하는 로직을 WinningTicket 에게 위임

yxxnghwan committed:
style(all): 개행, 공백, 메서드 순서 정렬, 불필요한 import 삭제

yxxnghwan committed:
test: LottoNumberGenerateStrategy에 대한 테스트 케이스 작성

yxxnghwan committed:
feat: 로또 WinningPrize 전략 패턴 정의

yxxnghwan committed:
refactor: LottoGame이 WinningPrizeStrategy를 생성 파라미터로 입력받아 상금 도출 로직 사용

yxxnghwan committed:
refactor: WinningResultOutputView에서 WinningResultDto를 받아 정보를 출력하도록 변경

yxxnghwan committed:
refactor: DefaultLottoWinningPrizeStrategy에 대한 테스트 케이스 작성

yxxnghwan committed:
refactor: 긴 메서드 뽑기, 개행 조절

yxxnghwan committed:
refactor: 사용하지 않는 변수, 메서드 제거

jnsorn commented on src/main/java/view/InputView.java:
```diff
@@ -0,0 +1,10 @@
+    Scanner scanner = new Scanner(System.in);
+
+    T getUserInputData();
+}
```
> 이번 미션에서 단순히 InputView, OutputView 두 가지 View로 모든 UI로직을 담기엔 한 파일에 코드가 길어지고, 한 클래스가 여러 기능을 해야한다고 생각하게 되어서 각 기능마다 View 클래스를 따로 만들었는데요, View 클래스들의 공통적인 부분을 묶어 인터페이스로 만들었는데, 막상 인터페이스의 주요 기능인 다형성을 사용하지는 않았습니다. 이러한 구조의 경우 오버엔지니어링이었을까요?🤔

이 구조에 대해 이야기해보기 전에 알렉스가 생각하는 오버엔지니어링은 무엇인지가 궁금해요. 또 인터페이스를 사용하는 이유에는 어떤 것들이 있을까요?

yxxnghwan replied:
저는 어플리케이션에 꼭 필요하지 않은 과한 설계로 인해서 얻는 장점보다 설계의 복잡함으로 인한 단점이 더 커지면 오버엔지니어링이라고 생각합니다. 인터페이스는 다형성이나 OCP 외에도 특징이 겹치는 클래스를 군집하고 어떤 가이드를 제공하는 역할도 있긴 하지만, 이번에 View를 각각 나누면서 설계의 복잡함으로 인한 피로도가 꽤 있었어요...😂

jnsorn replied:
말씀해주신대로라면 인터페이스를 통해 view를 군집하고 가이드를 제공하는 장점이 있지만, 설계의 복잡함으로 인한 피로도가 증가하는 단점이 있는 구조네요. 개인적으로 저는 코드를 읽으면서 단점이 더 크게 다가왔기 때문에 단순한 InputView, OutputView 두 개를 이용해 구현할 것 같아요. :) 특히 클래스를 군집할 수 있는 장점은 현재 구조에서는 패키지로 충분히 표현할 수 있다고 생각합니다. 현재 새로운 view가 추가되면 사용하는 곳에서 구현클래스를 통해 메서드를 호출해야 하는 구조라 더욱 이 구조의 장점을 찾기가 어렵네요. 이런 설계와 같은 부분은 앞으로 미션을 진행하면서 지금처럼 다양하게 고민해보시면, 더 적절하고 유연한 구조를 설계하실 수 있을 거예요. 지금은 여러 고민을 해보고 다양한 방법을 시도하신 것 자체로도 너무 좋습니다.

jnsorn commented on src/main/java/model/WinningResultDto.java:
```diff
@@ -0,0 +1,31 @@
+package model;
+
+public class WinningResultDto {
```
> View에서 모델의 비즈니스 로직을 노출시키지 않으려고 DTO를 넘기는 것으로 알고 있습니다. 근데 로직을 가지지 않고 상수처럼 사용하는 enum도 DTO 써서 View에 전달하는게 맞는지 궁금합니다.(처음엔 DTO없이 View에 enum을 넘기도록 구현했었으나, enum이 가지는 정보를 축소하면서 DTO를 생성해서 넘기는 구조로 변경했습니다.)

저는 enum도 DTO를 이용해서 전달하는 것이 좋다고 생각해요. 로직을 가지지 않고 상수처럼 사용하는 enum이라고 말씀하셨지만, 요구사항이 변경되고 도메인 규칙이 변하면서 추후에 로직을 갖게될 가능성이 충분히 있다고 생각합니다. (ex. 등수에 따라 세금이 달라진다면? 상금에 수수료를 곱해 실수령액을 계산해야 한다면?) 특히, WinningResultDto는 view에서 필요한 형태로 가공해주는 DTO 역할을 충실히 하고 있는 것으로 보여요.😃

jnsorn commented on src/main/java/LottoMain.java:
```diff
@@ -0,0 +1,37 @@
+
+    private static LottoTickets purchaseLottoTickets() {
+        int purchaseMoney = controller.inputPurchaseMoney();
+        LottoTickets lottoTickets = controller.createLottoTickets(purchaseMoney, new LottoNumberGenerateStrategy());
```
> 전략패턴을 최초로 주입하는 위치가 궁금합니다. 현재는 메인메서드부터 전략패턴을 주입해서 메인메서드->컨트롤러->도메인 이렇게 전달하면서 전략 패턴을 사용중인데요, 어느 부분에서부터 전략 패턴을 정하고 파라미터를 넘겨주지 않아도 되는지 궁금합니다!

해당 질문은 2단계 미션을 진행하면 자연스럽게 답을 찾을 수 있을 것 같아요. 🙂 2단계에서 같이 얘기 나눠봐요!

jnsorn commented on src/main/java/LottoMain.java:
```diff
@@ -0,0 +1,37 @@
+import model.DefaultLottoWinningPrizeStrategy;
+import model.WinningPrize;
+
+public class LottoMain {
+    public static LottoController controller = new LottoController();
```
LottoMain과 LottoController의 역할은 무엇인가요?

yxxnghwan replied:
메인은 해당 어플리케이션의 시나리오, 컨트롤러는 그 시나리오에서 필요한 기능 요청을 중계하는 중계자라고 생각하고 구현을 했습니다.
단 하나의 시나리오만 있는 이런 단일 흐름의 어플리케이션에서 컨트롤러와 메인메서드의 역할을 어떻게 나누는게 정답인지는 아직 감이 하나도 안잡힙니다..😂
다른 크루들의 구현 코드는 메인 메서드에서 컨트롤러의 생성자와 run() 이라는 메서드 하나만 호출하도록 구현한 크루들도 있었는데, 그렇게 구조를 짜는게 더 나은 구조일까요..?

jnsorn replied:
물론 정답은 없지만 저라면 main 메서드에서는 controller를 호출하고 controller에서 로또 게임 시나리오를 수행해 나갈 것 같아요! 그 이유는 main 메서드와 controller가 각각 아래의 역할을 갖고 있다고 생각하기 때문이에요. 

- Main Application(main 메서드) : JAVA 어플리케이션을 실행하기 위한 용도로, 프로그램을 실행을 담당
- Controller(LottoController) : 요청에 대해 적절한 책임을 갖고 있는 객체(model) 호출 및 view와 model 간의 데이터를 전달

만약 이 프로그램이 인기가 많아져서 로또게임 외에도 다양한 게임을 제공하는 프로그램이 된다면? 자동차 경주에 대한 요청도 받고, 로또 게임, 카드 게임 등 다양한 요청을 받게 되면 그때는 controller와 main 메서드를 어떤 식으로 구성하는게 좋을까요? 여러 고민을 해보면서 각각의 장단점을 찾아보면 좋을 것 같아요. 🙂

jnsorn commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,82 @@
+    }
+
+    public Integer inputPurchaseMoney() {
+        return (new PurchaseMoneyInputView()).getUserInputData();
```
view를 계속 생성할 필요가 있을까요?

yxxnghwan replied:
View를 담아두고 제공하는 ViewContainer를 정의했는데요, 
View를 인터페이스를 구현하도록 설계한 것을 살려 ViewMap을 만들어서 Key로 원하는 View를 꺼낼 수 있게 구현하려는 중에 Generic을 사용한 것에 발목이 잡혔습니다 ㅠㅠ... 결국 그냥 ViewContainer에 각각 View들을 필드로 가지게 했는데요... 해당 구조에 문제가 있을까요..? View설계의 복잡함으로 자꾸 문제가 발생하니 그냥 다시 InputView, OutputView 두개의 View로 단순화시키고 싶은 생각이 듭니다 ㅠㅠ

jnsorn replied:
ViewContainer에서 각 view의 인스턴스를 가지고 있는 방법도 있지만, 각 View에서 인스턴스를 미리 생성해두고 필요 시 미리 생성해둔 인스턴스를 반환하는 방법도 있을 것 같아요. (싱글턴 패턴)

jnsorn commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,82 @@
+        List<LottoTicketDto> dto = lottoTickets.getTickets().stream()
+                .map(lottoTicket -> new LottoTicketDto(lottoTicket.lottoNumberValues()))
+                .collect(Collectors.toList());
+        (new LottoTicketOutputView()).printOutputData(dto);
```
컨벤션이 맞나요?

yxxnghwan replied:
수정했어요! 감사합니다!

jnsorn replied:
아직 고쳐지지 않은 곳이 있는 것 같아요! 👀

jnsorn commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,82 @@
+        (new WinningResultOutputView()).printOutputData(winningResultDtos);
+    }
+
+    private WinningResultDto toWinningResultDto(Entry<WinningPrize, Integer> entry) {
```
`entry`보다 더 명확한 이름을 지어주세요! (혹은 winningPrize와 Integer를 각각 받는 방법도 고민해봐요)

yxxnghwan replied:
각각 받는 방식으로 변경했습니다!

jnsorn commented on src/main/java/validator/LottoNumberValidator.java:
```diff
@@ -0,0 +1,26 @@
+        if (isNotCorrectNumber(parsedNumber)) {
+            throw new IllegalArgumentException(LOTTO_NUMBER_RANGE_ERROR_MESSAGE);
+        }
+    }
```
` 함수(또는 메서드)의 길이가 10라인을 넘어가지 않도록 구현한다.`

yxxnghwan replied:
validator가 파싱에 대한 내용을 아예 모르도록 설계를 변경하면서 해결되었어요! 원시 데이터에 대해서만 검증하도록 하고, 파싱에 대한 검증은 View가 하는게 맞다고 생각했어요. validator들은 유틸성 클래스이니 만큼 공통된 검증에 대한 책임만 갖는 것이 좋다고 생각하여 변경했습니다!

jnsorn commented on src/main/java/validator/LottoNumberValidator.java:
```diff
@@ -0,0 +1,26 @@
+        }
+    }
+
+    private static boolean isNotCorrectNumber(int number) {
```
NotCorrect는 너무 많은 의미를 담고있어요! 조금 더 구체적인 이름을 지어볼까요?

yxxnghwan replied:
해당 부분은 범위를 체크하는 로직만 있네요! isNotInRangeNumber로 메서드명 변경했습니다!

jnsorn commented on src/main/java/validator/PurchaseMoneyValidator.java:
```diff
@@ -0,0 +1,29 @@
+    private static final String NOT_INTEGER_ERROR_MESSAGE = "금액을 정수로 입력해주세요.";
+
+    public static void validate(String purchaseMoneyString) {
+        int purchaseMoney = checkInteger(purchaseMoneyString);
```
메서드명은 check지만, int로 변환해주는 메서드네요 🤔

yxxnghwan replied:
수정완료!

jnsorn commented on src/main/java/view/BonusNumberInputView.java:
```diff
@@ -0,0 +1,25 @@
+
+    private Integer inputBonusNumber() {
+        String inputData = scanner.nextLine();
+        LottoNumberValidator.validate(inputData.trim());
```
inputData가 null이 들어온다면?

yxxnghwan replied:
InputView를 구현하는 ConsoleInputView라는 추상 클래스를 정의해서 콘솔의 입력값을 검증하도록 로직을 추가했어요!
기존의 다른 InputView를 구현하던 클래스들은 ConsoleInputView를 상속하도록 변경했습니다.

jnsorn commented on src/main/java/view/LottoTicketOutputView.java:
```diff
@@ -0,0 +1,29 @@
+    }
+
+    private String toLottoTicketPrintForm(LottoTicketDto lottoTicket) {
+        return RIGHT_LIST_COVER + joiningLottoNumbers(lottoTicket) + LEFT_LIST_COVER;
```
String +연산, StringBuilder, StringBuffer 클래스에 대해 알아볼까요!

yxxnghwan replied:
String은 더할 때마다 객체가 새로 생성되는군요...
StringBuffer는 스레드 세이프이구요!
감사합니다 또링! 지적해주신 덕분에 공부할 수 있었어요!😁

jnsorn commented on src/main/java/view/PurchaseMoneyInputView.java:
```diff
@@ -0,0 +1,25 @@
+
+    private Integer inputPurchaseMoney() {
+        String userInput = scanner.nextLine();
+        PurchaseMoneyValidator.validate(userInput);
```
알렉스는 View에서 어디까지 검증해야한다고 생각하시나요?

yxxnghwan replied:
저는 View가 가장 엄격한 검증을 하도록 했는데요,
도메인에선 View에서 파싱되어 들어온 자바의 원시 데이터 기준으로만 검증하도록 했습니다.
View의 경우에 파싱의 검증에서 문제가 있든, 도메인 기준 검증에서 문제가 있든 잘못된 값을 입력시에 재입력을 받도록 하는 것이 UX적으로 맞는 설계라고 생각했어요. 또링은 어떤 의견이신가요?

jnsorn replied:
저도 알렉스와 비슷한 것 같아요:) view에서는 입력 형태(null, type)에 대한 검증을 하고, 도메인에서는 도메인 규칙을 검증하는 것을 선호합니다. 
view에서 도메인 규칙을 검증하게 된다면, 도메인 규칙이 여기저기 흩어져있게 될 것이고 그건 유지보수에 좋지 않다고 생각해요.
또, 반대로 view에서 입력 형태마저 검증하지 않는다면 모든 입력 형태에 대한 방어 로직이 여기 저기 흩어져 있는 형태가 될 것이기 때문에, 적절히 나눠서 검증해주는 것을 선호합니다.

jnsorn commented on src/main/java/model/WinningPrize.java:
```diff
@@ -0,0 +1,19 @@
+package model;
+
+public enum WinningPrize {
+    FIRST(2_000_000_000),
```
가독성이 좋아요! 👏🏻

jnsorn commented on src/main/java/model/LottoNumberGenerateStrategy.java:
```diff
@@ -0,0 +1,23 @@
+        List<Integer> numbers = generateRangerNumbers();
+        Collections.shuffle(numbers);
+        return numbers.stream()
+                .limit(6)
```
6이라는 숫자가 의미하는 바는 무엇일까요?

yxxnghwan replied:
상수로 뽑았습니다!

jnsorn commented on src/main/java/model/LottoNumberGenerateStrategy.java:
```diff
@@ -0,0 +1,23 @@
+    }
+
+    private List<Integer> generateRangerNumbers() {
+        return IntStream.rangeClosed(1, 45)
```
`1`, `45` 👀

yxxnghwan replied:
앗... 상수...😅

jnsorn commented on src/test/java/validator/LottoNumberValidatorTest.java:
```diff
@@ -0,0 +1,18 @@
+    @ValueSource(strings = {"ada", "46"})
+    void checkInvalidInput(String input) {
+        assertThatThrownBy(() -> LottoNumberValidator.validate(input))
+                .isInstanceOf(IllegalArgumentException.class);
```
에러메세지까지 비교해서 더 정확한 테스트를 해봅시다. 😎

yxxnghwan replied:
메시지 비교 추가했습니다!
가존 @ParameterizedTest 사용해서 작성됐던 코드는 부득이하게 나누게 되었어요.

jnsorn replied:
테스트 클래스에 프로덕션과 같은 상수를 선언하여 메세지 비교하신 것 확인했습니다. 하나의 public 메서드가 다양한 예외메세지를 보여줄 수 있는 경우엔 특히 더 메세지 비교가 중요한 것 같아요 :) 
참고차 말씀드리자면, 저는 테스트에서 에러메세지를 비교하는 경우 프로덕션의 상수를 그대로 사용합니다.(private이라면 default로 변경해서 사용) 그 이유는 테스트의 유지보수성 때문인데요, 에러메세지가 변경되면 테스트가 깨지고 결국 테스트의 에러메세지도 지속적으로 관리해줘야 하기 때문에 관리 포인트를 한 곳으로 모은다고 생각해주시면 될 것 같습니다.ㅎㅎ 이건 정말 취향의 차이라 참고만 해주세요!

jnsorn commented on src/main/java/model/LottoTicket.java:
```diff
@@ -0,0 +1,23 @@
+import java.util.stream.Collectors;
+
+public class LottoTicket {
+    private final List<LottoNumber> lottoNumbers;
```
로또 티켓의 로또 번호들이 중복될 수 없는 규칙과 아주 잘 맞는 자료구조가 있을 것 같아요 🧐

yxxnghwan replied:
Set! 변경했습니다!

jnsorn commented on src/main/java/model/LottoNumber.java:
```diff
@@ -0,0 +1,24 @@
+    private final int number;
+
+    public LottoNumber(int number) {
+        this.number = checkNumber(number);
```
LottoNumber를 검증하는 로직이 두 곳에 흩어져 있어요! LottoNumber의 checkNumber와 LottoNumberValidator는 어떤 차이가 있나요? 어떻게 한 곳으로 모을 수 있을까요?

yxxnghwan replied:
LottoNumberValidator가 유틸성 클래스이니 LottoNumber가 LottoNumberValidator의 기능을 이용해서 검증하도록 변경했어요!

jnsorn reviewed (changes requested):
안녕하세요, 알렉스! 리뷰어 또링입니다. 👋

코드가 깔끔하고, 프로그래밍 요구사항을 잘 지켜주셔서 읽기 편했습니다. 또, 커밋 로그를 보니 TDD를 잘 지켜서 구현해주신 것 같아요. 자동차미션에서 배운 내용도 잘 적용해주신 것 같아서 아주 좋습니다. 👍🏼
질문주신 내용에 대한 답변은 관련 코드 코멘트로 남겨두었어요. 제 의견을 그대로 받아들이기 보다는, 여러 고민 해보시고 알렉스의 생각을 코멘트로 남겨주시면 더 깊고 다양한 얘기를 할 수 있을 것 같습니다! 

혹시 궁금하거나 얘기 나누고 싶은 내용이 있다면 편하게 DM 주세요. 😃

yxxnghwan committed:
style: LottoController 컨벤션 수정

yxxnghwan committed:
refactor: LottoNumberValidator 메서드 분리, 메서드 네이밍 수정

yxxnghwan committed:
refactor: 메서드 분리, 리네이밍, 시그니처 변경, LottoTicket의 자료형 List -> Set으로 변경

yxxnghwan committed:
fix: 컴파일 에러 fix

yxxnghwan committed:
fix: 컴파일 에러 fix

yxxnghwan committed:
fix: 컴파일 에러 fix

yxxnghwan committed:
test: 검증 테스트에서 에러 발생 메시지까지 검증하도록 변경

yxxnghwan committed:
refactor: validator의 파싱에 대한 검증을 View에 넘김, View 구조 재설계

validator는 유틸성 클래스이기 때문에 원시 데이터에 대한 검증만 하도록 변경
View에서는 입력값을 파싱하는 책임을 지도록 변경
InputView의 경우 ConsoleInputView를 상속해서 구현하도록 변경
입력에 대한 공통적인 검증로직을 ConsoleInputView가 가지도록 설계

yxxnghwan committed:
refactor: WinningNumberInputView에서 입력된 값을 Set으로 반환하도록 변경

yxxnghwan committed:
refactor: View를 메서드 호출때마다 생성해서 반환하지 않고 컨테이너에 저장해두고 필요할 때 불러오기

yxxnghwan committed:
style: 컨벤션 수정, 불필요한 import문 제거

yxxnghwan committed:
refactor: StringBuilder, printf 사용해서 결과 출력하기

yxxnghwan committed:
refactor: 상수화 놓친 것들 작업, 안쓰는 변수 제거

yxxnghwan commented:
안녕하세요 또링! 피드백 너무너무 감사합니다!😊
새로운 것도 많이 알게 되고, 고민도 많이 할 수 있었어요.
피드백 달아주신 내용 반영해서 다시 PR 올립니다!

여전히 View설계와 Controller의 역할에 대한 답을 찾지 못했지만...ㅠㅠ
저렇게 코드를 작성한 이유는 댓글로 남겨놨습니다!
또링이 보고 조언해주시면 감사하겠습니다.(사정없이 패주셔도 괜찮아요..😅)

jnsorn commented on src/main/java/model/WinningTicket.java:
```diff
@@ -0,0 +1,50 @@
+        this.bonusNumber = new LottoNumber(bonusNumber);
+    }
+
+    public List<Integer> winningNumberValues() {
+        return winningNumbers.stream()
+                .map(LottoNumber::value)
+                .collect(Collectors.toList());
+    }
+
+    public Integer bonusNumberValue() {
+        return bonusNumber.value();
+    }
```
접근제어자, 메서드 순서 확인해주세요!

jnsorn commented on src/main/java/model/LottoNumberGenerateStrategy.java:
```diff
@@ -0,0 +1,29 @@
+                .collect(Collectors.toSet());
+    }
+
+    private List<Integer> generateRangerNumbers() {
+        return IntStream.rangeClosed(MIN_LOTTO_NUMBER, MAX_LOTTO_NUMBER)
+                .boxed()
+                .collect(Collectors.toList());
+    }
```
로또를 생성할 때마다 로또 번호도 같이 생성되네요! 로또 번호가 1~45의 숫자로만 구성된다는 점을 이용하여 미리 생성해두고 필요할 때 마다 생성해 둔 번호를 재 사용하면 어떨까요?(Hint. 캐싱에 대해 알아보아요!)

jnsorn commented on src/main/java/controller/LottoController.java:
```diff
@@ -0,0 +1,93 @@
+        lottoGame = new LottoGame(lottoTickets, winningNumbers, bonusNumber, winningPrizeStrategy);
+    }
+
+    public Map<WinningPrize, Integer> winningResults() {
+        return lottoGame.winningResults();
+    }
+
+    public Double rateOfReturn() {
+        return lottoGame.getLottoRateOfReturn();
+    }
```
> 5.2.3 함수 이름
> 함수 이름은 lowerCamelCase 이다.
> 
> 함수 이름은 전형적으로 동사 혹은 동사 구이다. 예를들어, sendMessage 나 stop이다.

이 메서드들 뿐만 아니라 전반적으로 메서드 네이밍 컨벤션 확인해주세요!

jnsorn commented on src/test/java/validator/LottoNumberValidatorTest.java:
```diff
@@ -0,0 +1,20 @@
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
+
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
```
사용하지 않는 import는 제거해주세요!

jnsorn commented on src/test/java/model/LottoNumberGenerateStrategyTest.java:
```diff
@@ -0,0 +1,28 @@
+import org.junit.jupiter.api.DisplayName;
+import org.junit.jupiter.api.Test;
+
+class LottoNumberGenerateStrategyTest {
```
생성되는 숫자의 개수도 확인해봅시다!

jnsorn commented on src/main/java/model/LottoGame.java:
```diff
@@ -0,0 +1,82 @@
+import java.util.Map;
+import java.util.Set;
+
+public class LottoGame {
```
로또게임에서 게임 결과를 초기화하고, 게임 결과를 구하고, 수익률을 구하는 등 다양한 행위를 하고 있어요. 게임 결과를 담당하는 객체를 만들어볼까요?

jnsorn reviewed (approved):
안녕하세요, 알렉스! 리뷰 반영해주신 부분들 확인했어요. 👍🏼
이번에도 많은 걸 고민하고, 공부하신 것 같아요. 특히, 알렉스의 생각을 들을 수 있어서 좋았습니다!ㅎㅎ
1단계 미션은 여기서 마무리해도 될 것 같아서 이만 머지하도록 할게요.
2단계 진행하면서 더 많은 이야기 나눠봐요! 고생하셨습니다. 👏🏻
